[Code] Tispark support insert using SQL (#2471)
pingcap/tispark: docs/insert_sql_userguide.md

      setIfMissing("spark.app.name", getClass.getName).
      setIfMissing("spark.sql.extensions", "org.apache.spark.sql.TiExtensions").
      setIfMissing("spark.tispark.pd.addresses", "pd0:2379").
      setIfMissing("spark.tispark.tidb.addr", "tidb").
      setIfMissing("spark.tispark.tidb.port", "4000")

    val spark = SparkSession.builder.config(sparkConf).getOrCreate()
    ```

2. Write using Spark SQL.

    ```scala
    /* create table before run the code
    CREATE TABLE tpch_test.target_table_orders (
      `O_ORDERKEY` int(11) NOT NULL,
      `O_CUSTKEY` int(11) NOT NULL,
      `O_ORDERSTATUS` char(1) NOT NULL,
      `O_TOTALPRICE` decimal(15,2) NOT NULL,
      `O_ORDERDATE` date NOT NULL,
      `O_ORDERPRIORITY` char(15) NOT NULL,
      `O_CLERK` char(15) NOT NULL,
      `O_SHIPPRIORITY` int(11) NOT NULL,
      `O_COMMENT` varchar(79) NOT NULL
    )
    */
    
   // write data to tidb
    spark.sql(
          "insert into tpch_test.target_table_orders values" +
            "(1,1,'0',11.11,date'2022-1-1','first','spark',12345,'no comment')")

    // select data to write
    val df = spark.sql("select * from tpch_test.ORDERS")
   ```

## TiDB Options

> The following table shows the TiDB-specific options, which can be passed in through `TiDBOptions` or `SparkConf`.

Reviewer: Spark SQL can't accept TiDBOptions. So, we need not mention about`TiDBOptions` here?

| tidb.user                   | false    | -                      | TiDB User, needed when enableUpdateTableStatistics is true                                                               |
| tidb.password               | false    | -                      | TiDB Password, needed when enableUpdateTableStatistics is true                                                           |
| enableRegionSplit           | false    | true                   | To split Region to avoid hot Region during insertion                                                                     |
| scatterWaitMS               | false    | 300000                 | Max time to wait scatter region                                                                                          |
| writeThreadPerTask          | false    | 1                      | Thread number each spark task use to write data to TiKV                                                                  |
| bytesPerRegion              | false    | 100663296 (96M)        | Decrease this parameter to split more regions (increase write concurrency)                                               |

> | enableUpdateTableStatistics | false    | false                  | Update statistics in table `mysql.stats_meta` (`tidb.user` must own update privilege to table `mysql.stats_meta` if set true) |

Reviewer: - I think you can't pass enableRegionSplit,writeThreadPerTask,bytesPerRegion,scatterWaitMS beacause you don't get them in `newWriteBuilder`.
- database , table and pd.addresses is needn't.
- lack TiDB_ROW_FORMAT_VERSION

Author: I have fix point two and three. For point one, we will get them in  newWriteBuilder now.

## TiDB Version

TiDB's version must be **4.0.0 or later**.

## Limitations
TiSpark support write into partition table using INSERT SQL now.

But TiSpark does not support insert with a partition spec and a column list, like:

`INSERT INTO test.test PARTITION (i = 1)(s, k) VALUES ('hello', 'world');`

## Type Conversion for Write

The following types of SparkSQL Data are currently not supported for writing to TiDB:

- ArrayType
- MapType
- StructType

> Spark SQL and TIDB have different conversion rules. We follow the rule of TIDB when using DataSource API,

Reviewer 2: Plz give an example of how to cast such as String2TimeStamp and make spark link

> Spark SQL and TIDB have different conversion rules. We follow the rule of TIDB when using DataSource API,

Reviewer: I think DataSource API may not follow TiDB

> but follow the rule of Spark SQL when using it.

Reviewer: ```suggestion
but follow the rule of Spark SQL when using insert.
```

It supports insert data using Spark SQL now.

## Use Spark SQL
1. Initiate `SparkConf`.

    ```scala
    val sparkConf = new SparkConf().
      setIfMissing("spark.master", "local[*]").
      setIfMissing("spark.app.name", getClass.getName).
      setIfMissing("spark.sql.extensions", "org.apache.spark.sql.TiExtensions").
      setIfMissing("spark.tispark.pd.addresses", "pd0:2379").
      setIfMissing("spark.tispark.tidb.addr", "tidb").
      setIfMissing("spark.tispark.tidb.port", "4000")

    val spark = SparkSession.builder.config(sparkConf).getOrCreate()
    ```

2. Write using Spark SQL.

    ```scala
    /* create table before run the code
    CREATE TABLE tpch_test.target_table_orders (
      `O_ORDERKEY` int(11) NOT NULL,
      `O_CUSTKEY` int(11) NOT NULL,
      `O_ORDERSTATUS` char(1) NOT NULL,
      `O_TOTALPRICE` decimal(15,2) NOT NULL,
      `O_ORDERDATE` date NOT NULL,
      `O_ORDERPRIORITY` char(15) NOT NULL,
      `O_CLERK` char(15) NOT NULL,
      `O_SHIPPRIORITY` int(11) NOT NULL,
      `O_COMMENT` varchar(79) NOT NULL
    )
    */
    
   // write data to tidb

>     spark.sql(

Reviewer 3: we don't need catalog?

Author: I‘ll  add it.

|-----------------------------|----------|-----------------|-------------------------------------------------------------------------------------------------------------------------------|
| tidb.addr                   | false    | -               | TiDB address, needed when enableUpdateTableStatistics is true                                                                 |
| tidb.port                   | false    | -               | TiDB Port, needed when enableUpdateTableStatistics is true                                                                    |
| tidb.user                   | false    | -               | TiDB User, needed when enableUpdateTableStatistics is true                                                                    |
| tidb.password               | false    | -               | TiDB Password, needed when enableUpdateTableStatistics is true                                                                |
| enableRegionSplit           | false    | true            | To split Region to avoid hot Region during insertion                                                                          |

> | scatterWaitMS               | false    | 300000          | Max time to wait scatter region                                                                                               |

Reviewer 3: Why delete doesn't have so many options?

Author: This options are used in write. Some of them are used to create jdbc client to update statistics, some of them are used to do region splits. 

Delete doesn't do these things. So it doesn't need.

      setIfMissing("spark.master", "local[*]").
      setIfMissing("spark.app.name", getClass.getName).
      setIfMissing("spark.sql.extensions", "org.apache.spark.sql.TiExtensions").
      setIfMissing("spark.tispark.pd.addresses", "pd0:2379").
      setIfMissing("spark.tispark.tidb.addr", "tidb").
      setIfMissing("spark.tispark.tidb.port", "4000")

    val spark = SparkSession.builder.config(sparkConf).getOrCreate()
    ```

2. Write using Spark SQL.

    ```scala
    /* create table before run the code
    CREATE TABLE tpch_test.target_table_orders (
      `O_ORDERKEY` int(11) NOT NULL,
      `O_CUSTKEY` int(11) NOT NULL,
      `O_ORDERSTATUS` char(1) NOT NULL,
      `O_TOTALPRICE` decimal(15,2) NOT NULL,
      `O_ORDERDATE` date NOT NULL,
      `O_ORDERPRIORITY` char(15) NOT NULL,
      `O_CLERK` char(15) NOT NULL,
      `O_SHIPPRIORITY` int(11) NOT NULL,
      `O_COMMENT` varchar(79) NOT NULL
    )
    */
   
   //use catalog
   spark.sql("use tidb_catalog")
    
   // write data to tidb
    spark.sql(
          "insert into tpch_test.target_table_orders values" +
            "(1,1,'0',11.11,date'2022-1-1','first','spark',12345,'no comment')")

    // select data

>     val df = spark.sql("select * from tpch_test.target_table_orders")

Reviewer: ```suggestion
    spark.sql("select * from tpch_test.target_table_orders").show
```
use show to trigger an action,so that user can see the result

   ```

## TiDB Options

> The following table shows the TiDB-specific options, which can be passed in through `SparkConf`.

Reviewer: password can't be passed by SparkConf, it can only be passed by `spark.conf.set`. they are different for the latter one is session level config, and it will include the configs in SparkConf (`SparkContext.getOrCreate().getConf`)
you need to remind user not set password in SparkConf at least. about other configs, they can set by both the SparkConf and `spark.conf.set`

      setIfMissing("spark.sql.extensions", "org.apache.spark.sql.TiExtensions").
      setIfMissing("spark.tispark.pd.addresses", "pd0:2379").
      setIfMissing("spark.tispark.tidb.addr", "tidb").
      setIfMissing("spark.tispark.tidb.port", "4000")

    val spark = SparkSession.builder.config(sparkConf).getOrCreate()
    ```

2. Write using Spark SQL.

    ```scala
    /* create table before run the code
    CREATE TABLE tpch_test.target_table_orders (
      `O_ORDERKEY` int(11) NOT NULL,
      `O_CUSTKEY` int(11) NOT NULL,
      `O_ORDERSTATUS` char(1) NOT NULL,
      `O_TOTALPRICE` decimal(15,2) NOT NULL,
      `O_ORDERDATE` date NOT NULL,
      `O_ORDERPRIORITY` char(15) NOT NULL,
      `O_CLERK` char(15) NOT NULL,
      `O_SHIPPRIORITY` int(11) NOT NULL,
      `O_COMMENT` varchar(79) NOT NULL
    )
    */
   
   //use catalog
   spark.sql("use tidb_catalog")
    
   // write data to tidb
    spark.sql(
          "insert into tpch_test.target_table_orders values" +
            "(1,1,'0',11.11,date'2022-1-1','first','spark',12345,'no comment')")

    // select data
    spark.sql("select * from tpch_test.target_table_orders").show
   ```

> ## TiDB Options

Reviewer: I think options refer to datasource options. it is better to use configs or Configurations  in spark sql

> The following table shows the TiDB-specific options, which can be passed in through `SparkConf` or `spark.conf.set`.

Reviewer: ```suggestion
The following table shows the insert configurations, which can be passed in through `SparkConf` or `spark.conf.set`.
```
